File: server/src/detection/avoidance.test.ts
import { describe, expect, it } from "vitest";
import { hasRecentActivityOnOtherTasks } from "./avoidance.js";
import { makeActivityEvent, makeTask } from "./detection-test-fixtures.js";

describe("hasRecentActivityOnOtherTasks", () => {
  it("returns true when another task had a task_start within the window", () => {
    const topTask = makeTask({ id: 1 });
    const now = new Date("2026-07-05T10:00:00.000Z");
    const events = [
      makeActivityEvent({
        type: "task_start",
        task_id: 2,
        created_at: "2026-07-05T09:45:00.000Z",
      }),
    ];

    expect(hasRecentActivityOnOtherTasks(topTask, now, events, 30)).toBe(true);
  });

  it("returns true when another task had a task_update within the window", () => {
    const topTask = makeTask({ id: 1 });
    const now = new Date("2026-07-05T10:00:00.000Z");
    const events = [
      makeActivityEvent({
        type: "task_update",
        task_id: 2,
        created_at: "2026-07-05T09:45:00.000Z",
      }),
    ];

    expect(hasRecentActivityOnOtherTasks(topTask, now, events, 30)).toBe(true);
  });

  it("returns false when the activity is older than the window", () => {
    const topTask = makeTask({ id: 1 });
    const now = new Date("2026-07-05T10:00:00.000Z");
    const events = [
      makeActivityEvent({
        type: "task_start",
        task_id: 2,
        created_at: "2026-07-05T09:29:00.000Z",
      }),
    ];

    expect(hasRecentActivityOnOtherTasks(topTask, now, events, 30)).toBe(false);
  });

  it("returns false when the activity is on the top-priority task itself", () => {
    const topTask = makeTask({ id: 1 });
    const now = new Date("2026-07-05T10:00:00.000Z");
    const events = [
      makeActivityEvent({
        type: "task_start",
        task_id: 1,
        created_at: "2026-07-05T09:45:00.000Z",
      }),
    ];

    expect(hasRecentActivityOnOtherTasks(topTask, now, events, 30)).toBe(false);
  });

  it("returns false for activity types unrelated to task work (e.g. checkin)", () => {
    const topTask = makeTask({ id: 1 });
    const now = new Date("2026-07-05T10:00:00.000Z");
    const events = [
      makeActivityEvent({
        type: "checkin",
        task_id: 2,
        created_at: "2026-07-05T09:45:00.000Z",
      }),
    ];

    expect(hasRecentActivityOnOtherTasks(topTask, now, events, 30)).toBe(false);
  });

  it("returns false when the event timestamp is in the future (clock skew)", () => {
    const topTask = makeTask({ id: 1 });
    const now = new Date("2026-07-05T10:00:00.000Z");
    const events = [
      makeActivityEvent({
        type: "task_start",
        task_id: 2,
        created_at: "2026-07-05T10:05:00.000Z",
      }),
    ];

    expect(hasRecentActivityOnOtherTasks(topTask, now, events, 30)).toBe(false);
  });

  it("returns false when task_id is null", () => {
    const topTask = makeTask({ id: 1 });
    const now = new Date("2026-07-05T10:00:00.000Z");
    const events = [
      makeActivityEvent({
        type: "task_start",
        task_id: null,
        created_at: "2026-07-05T09:45:00.000Z",
      }),
    ];

    expect(hasRecentActivityOnOtherTasks(topTask, now, events, 30)).toBe(false);
  });
});

File: spikes/flutter/detection/test/avoidance_test.dart
import 'package:aiboss_detection/aiboss_detection.dart';
import 'package:test/test.dart';
import 'fixtures.dart';

void main() {
  group('hasRecentActivityOnOtherTasks', () {
    final now = d('2026-07-05T10:00:00.000Z');
    bool check(String type, int? taskId, String createdAt) =>
        hasRecentActivityOnOtherTasks(makeTask(id: 1), now, [makeActivityEvent(type: type, taskId: taskId, createdAt: createdAt)], 30);

    test('returns true when another task had a task_start within the window', () => expect(check('task_start', 2, '2026-07-05T09:45:00.000Z'), isTrue));
    test('returns true when another task had a task_update within the window', () => expect(check('task_update', 2, '2026-07-05T09:45:00.000Z'), isTrue));
    test('returns false when the activity is older than the window', () => expect(check('task_start', 2, '2026-07-05T09:29:00.000Z'), isFalse));
    test('returns false when the activity is on the top-priority task itself', () => expect(check('task_start', 1, '2026-07-05T09:45:00.000Z'), isFalse));
    test('returns false for activity types unrelated to task work (e.g. checkin)', () => expect(check('checkin', 2, '2026-07-05T09:45:00.000Z'), isFalse));
    test('returns false when the event timestamp is in the future (clock skew)', () => expect(check('task_start', 2, '2026-07-05T10:05:00.000Z'), isFalse));
    test('returns false when task_id is null', () => expect(check('task_start', null, '2026-07-05T09:45:00.000Z'), isFalse));
  });
}
